File: haskell/blog-server/src/Domain/UseCases/Works/Filter/Input.hs
module Domain.UseCases.Works.Filter.Input (FilterWorksInput (..)) where

import Data.Text (Text)
import Domain.Entities.Work (Work)

type SearchWord = Text

type TagName = Text

data FilterWorksInput = FilterWorksInput
  { _searchWord :: SearchWord,
    _tags :: [TagName],
    _filterWorks :: SearchWord -> [TagName] -> IO [Work]
  }

File: fsharp/src/WorksServer/UseCases/Works/Filter/Input.fs
module WorksServer.UseCases.Works.Filter.Input

open WorksServer.Entities.Work
open WorksServer.Values.Category
open WorksServer.Values.LimitNumber
open WorksServer.Values.Offset

type FilterWorks = Option<string> -> Option<Category> -> Offset -> LimitNumber<int> -> Result<Work seq, string>

type FilterInput =
    { filterWorks: FilterWorks
      search: Option<string>
      category: Option<string>
      offset: int
      limit: int }
